### DISCUSSION TITLE
Error: Element type is invalid. Received a promise that resolves to: undefined. Lazy element type must resolve to a class or function.

### DISCUSSION THREAD

**Discussion by rahatsayyed** [vercel/next.js#73455] [Help]
### Summary

**How to Handle Promises in a Provider Component in Next.js 14 Layout?**

I’m trying to use `LucidProvider` in my Next.js 14 layout to wrap all my components. Inside the provider, I initialize `Lucid` and update the state. However, `Lucid` returns a promise, which is causing the following error:  

```
Error: Element type is invalid. Received a promise that resolves to: undefined. Lazy element type must resolve to a class or function.
```

Here’s the initialization logic in `LucidProvider.tsx`:

```typescript
const blockfrost = new Blockfrost(BF_URL, BF_PID);
Lucid(blockfrost, NETWORK).then(setLucid).catch((error) => console.log(error));
```

When I comment out these lines, the error disappears. I’ve tried different approaches, including using default and named imports, as well as `dynamic`, but none of them resolved the issue. How can I properly pass the resolved promise as a provider in the Next.js 14 layout?

`LucidProvider.tsx`
```

'use client';

import React, { createContext, useContext, useEffect, useState } from "react";
import { Blockfrost, Lucid, LucidEvolution, Network } from "@lucid-evolution/lucid";

interface LucidContextType {
  lucid: LucidEvolution | undefined;
}

const LucidContext = createContext<LucidContextType | undefined>(undefined);

const BF_URL = process.env.NEXT_PUBLIC_BF_URL!;
const BF_PID = process.env.NEXT_PUBLIC_BF_PID!;
const NETWORK = process.env.NEXT_PUBLIC_CARDANO_NETWORK as Network;

export interface ProvidersProps {
  children: React.ReactNode;
}
export default function LucidProvider({ children }: ProvidersProps) {
  const [lucid, setLucid] = useState<LucidEvolution>();

  useEffect(() => {
    const blockfrost = new Blockfrost(BF_URL, BF_PID);
    Lucid(blockfrost, NETWORK).then(setLucid).catch((error) => console.log(error));
  }, []);

  return (
    <LucidContext.Provider value={{ lucid, }}>
      {children}
    </LucidContext.Provider>
  );
};

export const useLucid = () => {
  const context = useContext(LucidContext);
  if (!context) throw new Error("useLucid must be used within a LucidProvider");
  return context;
};

```

`Layout.tsx`
```
import dynamic from "next/dynamic";
// const LucidProvider = dynamic(() => import('./LucidProvider'))
import LucidProvider from "./LucidProvider";
..
..
..
..
..
export default function RootLayout({ children }: { children: React.ReactNode }) {
  return (
    <html suppressHydrationWarning lang="en">
      <head />
      <body className={clsx(
        "min-h-screen bg-background font-sans antialiased",
        fontSans.variable,
        fontComfortaa.variable
      )}>
        <LucidProvider>
        <Navbar />
          <Providers themeProps={{ attribute: "class", defaultTheme: "dark" }}>
            <div className="relative flex flex-col h-screen">
              <main className="container mx-auto  flex-grow">{children}</main>
            </div>
          </Providers>
        </LucidProvider>
      </body>
    </html>
  );
}
```

### Additional information

In the below example i have removed the imports to keep it error free.
if you're recreating issue. you will have to import `@/app/LucidProvider` into `@/app/Layout`

### Example

https://github.com/youreachedrahat/Aiken-Lucid-Template/

**Comment by icyJoseph** [comment]
Hi,

I am not familiar with this library, but does it support a browser environment at all?

**Reply by rahatsayyed** [reply]
by browser environment you mean running on browser. YES.

this library is somewhat similar to `web3.js` and I am using it other way temporary.

example 1: [Usage]( https://github.com/youreachedrahat/Aiken-Lucid-Template/blob/5cfe3eb20bf82f9b08c4f50f4fceb827786652a9/client/components/WalletConnector/WalletConnectors.tsx#L11) && [declaration](https://github.com/youreachedrahat/Aiken-Lucid-Template/blob/main/client/config/lucid.ts)

example 2: [Usage & Declaration](https://github.com/youreachedrahat/aiken_minting_validator/blob/f1516588047636e94993c3a570786b18f4825fc5/offchain/app/Home.tsx#L15)


in both the cases I have to wrap my component in this code
```
"use client";

import dynamic from "next/dynamic";

const ComponentUsingLucid = dynamic(() => import("@/app/Home"), { ssr: false });

export default function Client() {
  return <ComponentUsingLucid />;
}
```

---
**P.S example 2 is from different repo**

**Reply by icyJoseph** [reply]
Mmm odd indeed, does it work with SSR set to false on the dynamic import?

Otherwise, could you make a small stackblitz example? Just the code exactly needed to make it fail. Just type, next.new in your browser search bar.

**Reply by rahatsayyed** [reply]
link for StackBlitz: [https://stackblitz.com/edit/lucidprovidererror](https://stackblitz.com/edit/lucidprovidererror?file=app%2Flayout.tsx)

Changes made
added dependency `"@lucid-evolution/lucid"`
added `LucidProvider.tsx`
did  context wrapping in `Layout.tsx`

**Reply by rahatsayyed** [reply]
hey @icyJoseph do you have any update?

**Reply by icyJoseph** [reply]
Hey, no sorry, I've been playing around with it, and yeah I get the error, but also a bunch of other warnings and silent errors

Not really sure what's up with the library 🤷 

I think, now, you could setup a repository, and create an issue.

**Reply by rahatsayyed** [reply]
will try creating an issue, and as a temporary solution I will use states  

Thanks for the Help.
